### PR TITLE
[WIP] docs: Add rstcheck for generarated rst

### PR DESCRIPTION
**Pull request by phlax**
Commit Message: docs: Add rstcheck for generarated rst
Additional Description:

this is a follow on from #15766 to use [rstcheck](https://pypi.org/project/rstcheck/)

it will take quite a lot  more cleaning up so i was hoping it wouldnt block #15766 

Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by phlax** [comment]
annoying rstcheck doesnt pick up the original error of `ref:` so i think it will be necessary to include the grep

File: tools/code_format/requirements.txt [hunk 1]
```diff
@@ -1,12 +1,21 @@
+docutils==0.16 \
```
Comment (phlax) [inline tools/code_format/requirements.txt:1]: remove - and revert unrelated changes in this file

File: tools/code_format/rst_check.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+    this is copied from rstcheck to capture stderr
+
+    """
+    args = rstcheck.parse_args()
+
+    if not args.files:
+        return 0
+
+    with rstcheck.enable_sphinx_if_possible():
```
Comment (phlax) [inline tools/code_format/rst_check.py:29]: im wondering if its possible to use importlib to assert that sphinx is available...

File: tools/code_format/rst_check.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+            for (filename, _errors) in results:
+                for error in _errors:
+                    line_number = error[0]
+                    message = error[1]
+
+                    if not re.match(r'\([A-Z]+/[0-9]+\)', message):
+                        message = '(ERROR/3) ' + message
+
+                    rstcheck.output_message('{}:{}: {}'.format(
```
Comment (phlax) [inline tools/code_format/rst_check.py:49]: probs want to strip the path prefix from the filename

File: docs/build.sh [hunk 4]
```diff
@@ -171,7 +171,17 @@ rsync -av \
...
+bazel run "${BAZEL_BUILD_OPTIONS[@]}" //tools/code_format:rst_check "${GENERATED_RST_DIR}" || {
+    LINTFAIL=yes
+}
+
 # To speed up validate_fragment invocations in validating_code_block
 bazel build "${BAZEL_BUILD_OPTIONS[@]}" //tools/config_validation:validate_fragment
 
 sphinx-build -W --keep-going -b html "${GENERATED_RST_DIR}" "${DOCS_OUTPUT_DIR}"
+
+if [[ -n "$LINTFAIL" ]]; then
```
Comment (phlax) [inline docs/build.sh:185]: handle this better

File: tools/code_format/rst_check.py (hunk 2 above)
Comment (phlax) [reply tools/code_format/rst_check.py:29]: and just enable, ill check...

**Review by htuch — COMMENTED** [review]
Nice; looks good directionally, let me know when ready for a full review pass.

**Comment by phlax** [comment]
im also a bit concerned that rstcheck seems to be unmaintained, and also has kinda obvious bugs that have been kicking around for ~6 years (https://github.com/myint/rstcheck/issues/76) without being addressed

all the same its kinda useful - so i think rather include it and add workarounds

**Comment by phlax** [comment]
surveying other linting tools for rst (again) they are seemingly similar - ie barely maintained if at all

the nature of rst is that it is complex (ie embeddable code) - not sure what is best here.

File: tools/base/runner.py [hunk 5]
```diff
@@ -73,6 +74,15 @@ def add_arguments(self, parser: argparse.ArgumentParser) -> None:
         """Override this method to add custom arguments to the arg parser"""
         pass
 
+    def parallel(self, cmd, args, handle):
+        errors = 0
+        pool = multiprocessing.Pool(multiprocessing.cpu_count())
```
Comment (htuch) [inline tools/base/runner.py:79]: Use a context manager `with`?

File: tools/code_format/rst_check.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+# TODO(phlax): move to .rstcheck.cfg when available
+RSTCHECK_GREP_FAIL = (" ref:", "\\[\\#")
+RSTCHECK_CONFIG=".rstcheck.cfg"
+RSTCHECK_COMMAND = ("rstcheck", "-r")
+
+
+@contextmanager
+def sphinx_enabled():
+    """Register Sphinx directives and roles."""
+    srcdir = tempfile.mkdtemp()
```
Comment (htuch) [inline tools/code_format/rst_check.py:32]: Why not `with tempfile.TemporaryDirectory()`?

File: tools/code_format/rst_check.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    def grep_unwanted(self, check: str) -> None:
+        response = self.fork(["grep", "-nr", "--include", "\\*.rst"] + [check])
+        if not response.returncode:
+            self.error("grep", [f"grepping found bad '{check}': {path}" for path in resp.stdout.split("\n")])
+        else:
+            self.succeed("grep", [f"grepping found no errors for '{check}'"])
+
+    def _rstcheck_handle(self, results: list) -> None:
+        """Handle multiprocessed error results of rstchecks"""
+        for (filename, _errors) in results:
```
Comment (htuch) [inline tools/code_format/rst_check.py:142]: Nit: parens not needed here and below

File: tools/code_format/rst_check.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+import rstcheck
+import sphinx
+
+from tools.base import checker, utils
+
+
+# things we dont want to see in generated docs
+# TODO(phlax): move to .rstcheck.cfg when available
+RSTCHECK_GREP_FAIL = (" ref:", "\\[\\#")
+RSTCHECK_CONFIG=".rstcheck.cfg"
```
[the remark this answers is not in the document]
Comment (phlax) [reply tools/code_format/rst_check.py:25]: i think it does - flake8 might not because we dont enable most pep8 checks yet, but yapf isnt happy with it
